sys/net/application_layer/coap/gcoap.c

```diff
@@ -587,12 +587,13 @@ kernel_pid_t gcoap_init(void)
     _pid = thread_create(_msg_stack, sizeof(_msg_stack), THREAD_PRIORITY_MAIN - 1,
                             THREAD_CREATE_STACKTEST, _event_loop, NULL, "coap");
 
+    mutex_init(&_coap_state.lock);
     /* Blank lists so we know if an entry is available. */
     memset(&_coap_state.open_reqs[0], 0, sizeof(_coap_state.open_reqs));
     memset(&_coap_state.observers[0], 0, sizeof(_coap_state.observers));
     memset(&_coap_state.observe_memos[0], 0, sizeof(_coap_state.observe_memos));
     /* randomize initial value */
-    _coap_state.last_message_id = random_uint32() & 0xFFFF;
+    atomic_init(&_coap_state.next_message_id, (uint16_t)random_uint32());
```

Comment:
Yes, I read about atomic_init() as being unsafe, but agree that it's moot in this context. I did not think to use atomic_store(), but thanks for pointing that out. I also thought about ATOMIC_INIT() in the initializer, but especially with the randomization, it seemed simplest/best to use the current solution.